File: pcc/py_runtime/py/py_int_decimal.py
"""pcc-Python replacement for py_runtime/src/py_int_decimal.c."""

__pcc_runtime_port__ = True

from pcc.extern import extern, c_abi_export, c_int64, c_ptr
from pcc.py_runtime.py.py_abi_constants import (
    PYINTOBJECT_DIGITS_OFFSET,
    PYINTOBJECT_NDIGITS_OFFSET,
    PYINTOBJECT_SIGN_OFFSET,
)
from pcc.unsafe import (
    free,
    load_i8,
    load_i32,
    malloc,
    memmove,
    null,
    ptr_add,
    ptr_is_null,
    store_i8,
    store_i32,
)


py_bigint_alloc = extern("py_bigint_alloc", (c_int64,), c_ptr)


def _load_u32(obj, offset: int) -> int:
    v: int = load_i32(obj, offset)
    if v < 0:
        v = v + 4294967296
    return v


def _store_u32(obj, offset: int, value: int) -> None:
    store_i32(obj, offset, value)


def _normalize(b) -> None:
    ndigits: int = load_i32(b, PYINTOBJECT_NDIGITS_OFFSET)
    while ndigits > 0 and _load_u32(b, PYINTOBJECT_DIGITS_OFFSET + (ndigits - 1) * 4) == 0:
        ndigits = ndigits - 1
    store_i32(b, PYINTOBJECT_NDIGITS_OFFSET, ndigits)
    if ndigits == 0:
        store_i32(b, PYINTOBJECT_SIGN_OFFSET, 0)
    elif load_i32(b, PYINTOBJECT_SIGN_OFFSET) == 0:
        store_i32(b, PYINTOBJECT_SIGN_OFFSET, 1)


def _bigint_copy(a):
    ndigits: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET)
    r = py_bigint_alloc(ndigits)
    if ptr_is_null(r):
        return r
    store_i32(r, PYINTOBJECT_SIGN_OFFSET, load_i32(a, PYINTOBJECT_SIGN_OFFSET))
    i: int = 0
    while i < ndigits:
        store_i32(r, PYINTOBJECT_DIGITS_OFFSET + i * 4, load_i32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4))
        i = i + 1
    return r


def _divmod_small_inplace(a, divisor: int) -> int:
    rem: int = 0
    i: int = load_i32(a, PYINTOBJECT_NDIGITS_OFFSET) - 1
    while i >= 0:
        cur: int = rem * 4294967296 + _load_u32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4)
        _store_u32(a, PYINTOBJECT_DIGITS_OFFSET + i * 4, cur // divisor)
        rem = cur % divisor
        i = i - 1
    _normalize(a)
    return rem


def _write_digit(buf, pos: int, digit: int) -> int:
    pos = pos - 1
    store_i8(buf, pos, 48 + digit)
    return pos


@c_abi_export("py_bigint_to_cstr")
def py_bigint_to_cstr(b):
    if load_i32(b, PYINTOBJECT_SIGN_OFFSET) == 0:
        s = malloc(2)
        if ptr_is_null(s):
            return s
        store_i8(s, 0, 48)
        store_i8(s, 1, 0)
        return s

    ndigits: int = load_i32(b, PYINTOBJECT_NDIGITS_OFFSET)
    bufsz: int = ndigits * 10 + 2
    buf = malloc(bufsz)
    if ptr_is_null(buf):
        return buf

    tmp = _bigint_copy(b)
    if ptr_is_null(tmp):
        free(buf)
        return null()
    store_i32(tmp, PYINTOBJECT_SIGN_OFFSET, 1)

    pos: int = bufsz - 1
    store_i8(buf, pos, 0)

    while load_i32(tmp, PYINTOBJECT_NDIGITS_OFFSET) > 0:
        rem: int = _divmod_small_inplace(tmp, 1000000000)
        more: bool = load_i32(tmp, PYINTOBJECT_NDIGITS_OFFSET) > 0
        if not more:
            if rem == 0:
                pos = _write_digit(buf, pos, 0)
            else:
                while rem > 0:
                    pos = _write_digit(buf, pos, rem % 10)
                    rem = rem // 10
        else:
            k: int = 0
            while k < 9:
                pos = _write_digit(buf, pos, rem % 10)
                rem = rem // 10
                k = k + 1
    free(tmp)

    if load_i32(b, PYINTOBJECT_SIGN_OFFSET) < 0:
        pos = pos - 1
        store_i8(buf, pos, 45)

    memmove(buf, ptr_add(buf, pos), bufsz - pos)
    return buf


@c_abi_export("py_bigint_to_base_cstr")
def py_bigint_to_base_cstr(b, base: int, prefix_ch: int):
    # Full base-{2,8,16} string for a bignum: "[-]0<prefix_ch><digits>"
    # (lowercase a-f). Mirrors py_bigint_to_base_cstr in py_int_decimal.c:
    # repeated divmod by the small base, one digit per iteration.
    neg: int = 0
    if load_i32(b, PYINTOBJECT_SIGN_OFFSET) < 0:
        neg = 1
    ndigits: int = load_i32(b, PYINTOBJECT_NDIGITS_OFFSET)
    bufsz: int = ndigits * 32 + 8   # base 2 is widest: <=32 bits per limb
    buf = malloc(bufsz)
    if ptr_is_null(buf):
        return buf
    tmp = _bigint_copy(b)
    if ptr_is_null(tmp):
        free(buf)
        return null()
    store_i32(tmp, PYINTOBJECT_SIGN_OFFSET, 1)           # work on the magnitude (b is nonzero here)
    pos: int = bufsz - 1
    store_i8(buf, pos, 0)           # NUL
    done: int = 0
    while done == 0:
        rem: int = _divmod_small_inplace(tmp, base)
        pos = pos - 1
        ch: int = 48 + rem          # '0' + rem
        if rem >= 10:
            ch = 97 + rem - 10      # 'a' + (rem - 10)
        store_i8(buf, pos, ch)
        if load_i32(tmp, PYINTOBJECT_NDIGITS_OFFSET) == 0:
            done = 1
    free(tmp)
    pos = pos - 1
    store_i8(buf, pos, prefix_ch)
    pos = pos - 1
    store_i8(buf, pos, 48)          # '0'
    if neg != 0:
        pos = pos - 1
        store_i8(buf, pos, 45)      # '-'
    memmove(buf, ptr_add(buf, pos), bufsz - pos)
    return buf


@c_abi_export("py_bigint_from_cstr")
def py_bigint_from_cstr(s):
    if ptr_is_null(s):
        return null()
    p = s
    sign: int = 1
    ch: int = load_i8(p, 0)
    if ch == 43:
        p = ptr_add(p, 1)
    elif ch == 45:
        sign = -1
        p = ptr_add(p, 1)
    if load_i8(p, 0) == 0:
        return null()

    acc = py_bigint_alloc(0)
    if ptr_is_null(acc):
        return acc

    while load_i8(p, 0) != 0:
        chunk: int = 0
        mul: int = 1
        count: int = 0
        ch = load_i8(p, 0)
        while ch != 0 and count < 9:
            if ch < 48 or ch > 57:
                free(acc)
                return null()
            chunk = chunk * 10 + (ch - 48)
            mul = mul * 10
            count = count + 1
            p = ptr_add(p, 1)
            ch = load_i8(p, 0)
        if count == 0:
            free(acc)
            return null()

        la: int = load_i32(acc, PYINTOBJECT_NDIGITS_OFFSET)
        nxt = py_bigint_alloc(la + 1)
        if ptr_is_null(nxt):
            free(acc)
            return null()

        carry: int = 0
        i: int = 0
        while i < la:
            cur: int = _load_u32(acc, PYINTOBJECT_DIGITS_OFFSET + i * 4) * mul + carry
            _store_u32(nxt, PYINTOBJECT_DIGITS_OFFSET + i * 4, cur & 4294967295)
            carry = cur >> 32
            i = i + 1
        _store_u32(nxt, PYINTOBJECT_DIGITS_OFFSET + la * 4, carry)

        carry = chunk
        i = 0
        while i < load_i32(nxt, PYINTOBJECT_NDIGITS_OFFSET) and carry != 0:
            cur = _load_u32(nxt, PYINTOBJECT_DIGITS_OFFSET + i * 4) + carry
            _store_u32(nxt, PYINTOBJECT_DIGITS_OFFSET + i * 4, cur & 4294967295)
            carry = cur >> 32
            i = i + 1

        store_i32(nxt, PYINTOBJECT_SIGN_OFFSET, 1)
        _normalize(nxt)
        free(acc)
        acc = nxt

    if load_i32(acc, PYINTOBJECT_NDIGITS_OFFSET) == 0:
        store_i32(acc, PYINTOBJECT_SIGN_OFFSET, 0)
    else:
        store_i32(acc, PYINTOBJECT_SIGN_OFFSET, sign)
    return acc

File: pcc/py_runtime/src/py_int_decimal.c
/* pcc/py_runtime/src/py_int_decimal.c
 *
 * Decimal string conversion split from py_int.c so the pcc-Python
 * runtime can replace it independently from core bignum arithmetic.
 */

#include "py_internal.h"
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>

static void py_bigint_normalize(PyIntObject *b) {
    while (b->ndigits > 0 && b->digits[b->ndigits - 1] == 0) {
        b->ndigits--;
    }
    if (b->ndigits == 0) b->sign = 0;
    else if (b->sign == 0) b->sign = 1;
}

static PyIntObject *bigint_copy(const PyIntObject *a) {
    PyIntObject *r = py_bigint_alloc(a->ndigits);
    if (r == NULL) return NULL;
    r->sign = a->sign;
    for (int32_t i = 0; i < a->ndigits; i++) r->digits[i] = a->digits[i];
    return r;
}

static uint32_t bigint_divmod_small_inplace(PyIntObject *a, uint32_t divisor) {
    uint64_t rem = 0;
    for (int32_t i = a->ndigits - 1; i >= 0; i--) {
        uint64_t cur = (rem << 32) | (uint64_t)a->digits[i];
        a->digits[i] = (uint32_t)(cur / divisor);
        rem = cur % divisor;
    }
    py_bigint_normalize(a);
    return (uint32_t)rem;
}

char *py_bigint_to_cstr(const PyIntObject *b) {
    if (b->sign == 0) {
        char *s = (char *)malloc(2);
        if (!s) return NULL;
        s[0] = '0'; s[1] = '\0';
        return s;
    }
    /* Upper bound: 10 decimal digits per 32-bit digit + sign + NUL. */
    size_t bufsz = (size_t)b->ndigits * 10 + 2;
    char *buf = (char *)malloc(bufsz);
    if (buf == NULL) return NULL;

    /* Work on a mutable copy. */
    PyIntObject *tmp = bigint_copy(b);
    if (tmp == NULL) { free(buf); return NULL; }
    tmp->sign = 1;  /* we know original is nonzero */

    size_t pos = bufsz;
    buf[--pos] = '\0';

    /* Repeatedly divmod by 10^9 to extract 9 decimal digits per iter. */
    const uint32_t CHUNK = 1000000000u;
    while (tmp->ndigits > 0) {
        uint32_t rem = bigint_divmod_small_inplace(tmp, CHUNK);
        bool more = tmp->ndigits > 0;
        /* Write exactly 9 digits if more chunks remain (zero-pad); else the
         * minimal representation. */
        int count = 0;
        if (!more) {
            /* strip leading zeros for the highest chunk */
            if (rem == 0) {
                buf[--pos] = '0';
                count = 1;
            } else {
                while (rem > 0) {
                    buf[--pos] = (char)('0' + (rem % 10));
                    rem /= 10;
                    count++;
                }
            }
        } else {
            for (int k = 0; k < 9; k++) {
                buf[--pos] = (char)('0' + (rem % 10));
                rem /= 10;
            }
            count = 9;
        }
        (void)count;
    }
    free(tmp);

    if (b->sign < 0) buf[--pos] = '-';
    /* Shift the string down so it starts at buf[0]. */
    size_t len = bufsz - pos;
    memmove(buf, buf + pos, len);
    /* Shrink allocation? Leave it — caller frees it regardless. */
    return buf;
}

/* Full base-{2,8,16} string for a bignum: "[-]0<prefix_ch><digits>" (lowercase
 * a-f for hex), e.g. py_bigint_to_base_cstr(B, 16, 'x') -> "0x10000...".
 * Mirrors py_bigint_to_cstr but divides by the (small) base, one digit per
 * iteration. Caller frees the returned malloc'd, NUL-terminated string. */
char *py_bigint_to_base_cstr(const PyIntObject *b, unsigned base, char prefix_ch) {
    int neg = b->sign < 0;
    /* base 2 is the widest: up to 32 binary digits per 32-bit limb, plus
     * sign + "0x" + NUL. */
    size_t bufsz = (size_t)b->ndigits * 32 + 8;
    char *buf = (char *)malloc(bufsz);
    if (buf == NULL) return NULL;
    PyIntObject *tmp = bigint_copy(b);
    if (tmp == NULL) { free(buf); return NULL; }
    tmp->sign = 1;  /* work on the magnitude (b is nonzero here) */
    size_t pos = bufsz;
    buf[--pos] = '\0';
    do {
        uint32_t rem = bigint_divmod_small_inplace(tmp, base);
        buf[--pos] = (rem < 10) ? (char)('0' + rem)
                                : (char)('a' + (rem - 10));
    } while (tmp->ndigits > 0);
    free(tmp);
    buf[--pos] = prefix_ch;
    buf[--pos] = '0';
    if (neg) buf[--pos] = '-';
    size_t len = bufsz - pos;  /* includes the NUL */
    memmove(buf, buf + pos, len);
    return buf;
}

/* Parse a decimal string (optional leading sign, then digits). Returns a
 * new bignum (never a tagged form) or NULL on error. */
PyIntObject *py_bigint_from_cstr(const char *s) {
    if (s == NULL) return NULL;
    const char *p = s;
    int sign = 1;
    if (*p == '+') p++;
    else if (*p == '-') { sign = -1; p++; }
    if (*p == '\0') return NULL;
    /* Accumulate by 10^9 chunks. */
    PyIntObject *acc = py_bigint_alloc(0);
    if (acc == NULL) return NULL;

    while (*p) {
        /* Read up to 9 decimal digits. */
        uint32_t chunk = 0;
        uint32_t mul = 1;
        int count = 0;
        while (*p && count < 9) {
            if (*p < '0' || *p > '9') { free(acc); return NULL; }
            chunk = chunk * 10 + (uint32_t)(*p - '0');
            mul *= 10;
            count++;
            p++;
        }
        if (count == 0) { free(acc); return NULL; }

        /* acc = acc * mul + chunk */
        /* Multiply in-place: allocate result with one extra digit. */
        int32_t la = acc->ndigits;
        PyIntObject *next = py_bigint_alloc(la + 1);
        if (next == NULL) { free(acc); return NULL; }
        uint64_t carry = 0;
        for (int32_t i = 0; i < la; i++) {
            uint64_t cur = (uint64_t)acc->digits[i] * (uint64_t)mul + carry;
            next->digits[i] = (uint32_t)(cur & 0xFFFFFFFFu);
            carry = cur >> 32;
        }
        next->digits[la] = (uint32_t)carry;
        /* Add chunk. */
        carry = chunk;
        for (int32_t i = 0; i < next->ndigits && carry; i++) {
            uint64_t cur = (uint64_t)next->digits[i] + carry;
            next->digits[i] = (uint32_t)(cur & 0xFFFFFFFFu);
            carry = cur >> 32;
        }
        next->sign = 1;
        py_bigint_normalize(next);
        free(acc);
        acc = next;
    }
    if (acc->ndigits == 0) {
        acc->sign = 0;
    } else {
        acc->sign = sign;
    }
    return acc;
}
